src/tike/ptycho/exitwave.py

```diff
@@ -0,0 +1,180 @@
+    """We use an iterative/recursive method for finding the steplengths, 
+    and this is what we use as initialization for that method."""
+
+    unmeasured_pixels_scaling: float = 1.00
+    """Depending on how we control scaling for the exitwaves, we might need to scale up or down 
+    the number of photons in the unmeasured regions for the exitwave updates in Fourier space."""
+
+    measured_pixels: typing.Union[npt.NDArray, None] = dataclasses.field(
+        default_factory=lambda: np.ones(1, dtype=bool))
+    """A binary array that defines spatial regions on the detector where we have measured pixels."""
+
+    def copy_to_device(self, comm):
+        """Copy to the current GPU memory."""
+        if self.measured_pixels is not None:
+            self.measured_pixels = cp.asarray(self.measured_pixels)
+        # if self.measured_pixels is not None:
+        #     self.measured_pixels = comm.pool.bcast([self.measured_pixels])
+        return self
+
+    def copy_to_host(self):
+        """Copy to the host CPU memory."""
+        if self.measured_pixels is not None:
+            self.measured_pixels = cp.asnumpy(self.measured_pixels)
+        # if self.measured_pixels is not None:
+        #     self.measured_pixels = cp.asnumpy( self.measured_pixels[0] )
+        return self
+
+    def resample(self, factor: float) -> ExitWaveOptions:
+        """Return a new `ExitWaveOptions` with the parameters rescaled."""
+        return ExitWaveOptions(
+            noise_model=self.noise_model,
+            step_length_weight=self.step_length_weight,
+            step_length_start=self.step_length_start,
+            step_length_usemodes=self.step_length_usemodes,
+            measured_pixels=crop_fourier_space(
+                self.measured_pixels,
+                int(self.measured_pixels.shape[-1] * factor),
+            ),
+            unmeasured_pixels_scaling=self.unmeasured_pixels_scaling)
+
+
+def poisson_steplength_all_modes(
+    xi,
+    abs2_Psi,
+    I_e,
+    I_m,
+    measured_pixels,
+    step_length,
+    weight_avg,
+):
+    """Compute the optimal steplength for each exitwave mode independently.
+
+    Parameters
+    ----------
+    xi              : ND array ( # scan positions, # rows, # columns ), xi = 1 - I_m / I_e
+    abs2_Psi        : ND array ( # shared probe modes, None, # scan positions, # rows, # columns ), the squared absolute value of the calulated exitwaves       
+    I_m             : ND array ( # scan positions, # rows, # columns ), measured diffraction intensity
+    I_e             : ND array ( # scan positions, # rows, # columns ), calculated diffraction intensity 
+    measured_pixels : 2D binary array ( # rows, # columns ), the regions on the detector where we have defined measurements
+    step_length     : 2D array ( # shared probe modes, # scan positions), the steplength initializations
+    weight_avg      : float ( 0.0 <= weight_avg <= 1.0  ), the weight we use when computing a weighted average
+    """
```

Comment:
Look an any of the pre-existing function docstrings in Tike and format yours in the same way. The docstrings need to follow a specific syntax because they are converted into our documentation website automatically.

Reply:
Am doing this right now, more to come...

Reply:
I think I'm moving in the right direction here with the newest commit, let me know if not...